### PR TITLE
Feature: shield-charge improvements

### PR DESCRIPTION
**Pull request by eheimburg**
The tank character's special ability was lacking in several ways, which are addressed by this PR:
- it was using placeholder graphics
- it wasn't practical to actually use
- it wasn't obvious whether it was working or not
- the UI buttons didn't actually let you "charge up" the attack (you had to use the keyboard keys)

**Changes**
- The "charge up" time was increased from 1 to 2 seconds (that's `ExecTimeSeconds`)
- The "buffed" time was increased from 1 to 5 seconds (that's `EffectDurationSeconds`). On paper this seems like an insanely long time to be invincible, but in practice it feels fine: you can't sustain infinite invincibility because getting hit cancels the charge-up. So this really just gives you enough time to get into the fray before it wears off.
- Separated the graphics into two parts: "charging up" and "invincibility shield". The latter is only shown if the player reaches 100% charge (because that's the big obvious pay-off for the action). 
- When fully charged, the character does not animate hit-reacts (since they are not taking any damage!)

**Implementation Notes**
- There are two charge-up abilities and previously they shared the same client-side `ChargedActionFX` class. This is no longer practical, so I separated them into `ChargedLaunchProjectileActionFX` and `ChargedShieldActionFX`.
- I first tried to make `ChargedShieldActionFX` a subclass the old `ChargedActionFX` class, but the differences are very subtle and it became too hard to understand the flow. But I realized the biggest chunk of code in the old class could be turned into a reusable function, so I did that, and applied it to each ActionFX class that needed it.
- The client-side "stopped charging up" RPC now has a float that indicates how charged up the attack is.
- Since I was in the file already, I removed some cruft from `ChargedLaunchProjectileActionFX`: we didn't parent the old placeholder particles because rotating them looked weird, but the real ones look best parented.
- After fiddling with some approaches, decided the most elegant way to handle the UI button down/up events was to just glue that functionality into the existing `UIHUDButton` class. (My first approach was to use a separate component, but it ended up being messy to get the events piped around. Can still use that approach if this one feels icky, though.)
- Fixed edge-case bugs with `SpecialFXGraphic`

*Changes to CharacterSetController (in case it needs to be recreated after merge):*
- Unchecked "Has Exit Time" for transition from "Nothing" ->"Tank Shield Buff (Start)"
-  Added new Integer parameter "Invincible"
- On the HitReacts layer, Added a new node "Invincible (no reaction)". It has no animation (and only exists to let us reset the HitReact1 trigger)
- Added transition from "Nothing"->"Invincible (no reaction)" when HitReact1 is triggered AND Invincible > 0
- Changed transition from "Nothing"->"HitReact1" so that it requires both HitReact1 trigger AND Invincible == 0

*Changes to Hero Action Bar prefab:*
- removed the OnClick event handlers from all four buttons; this is now done from code (so we can subscribe to mouse-down and mouse-up events instead of just "click" events)

### CODE REVIEW

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.samples.coop?pullRequest=251) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.samples.coop?pullRequest=251) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.multiplayer.samples.coop?pullRequest=251) it.</sub>

**Review by fernando-cortez — COMMENTED** [review]
Small naming fix but looks good otherwise.

File: Assets/BossRoom/Scripts/Server/Game/Action/ChargedShieldAction.cs [hunk 1]
```diff
@@ -74,12 +74,25 @@ public override bool ShouldBecomeNonBlocking()
...
+            return Mathf.Clamp01(timeSpentChargingUp / Description.ExecTimeSeconds);
+        }
+
         public override void BuffValue(BuffableValue buffType, ref float buffedValue)
         {
             if (buffType == BuffableValue.PercentDamageReceived)
             {
-                float timeSpentChargingUp = m_StoppedChargingUpTime - TimeStarted;
-                float pctChargedUp = Mathf.Clamp01(timeSpentChargingUp / Description.ExecTimeSeconds);
+                float pctChargedUp = GetPercentChargedUp();
```
[the remark this answers is not in the document]
Comment (eheimburg) [reply Assets/BossRoom/Scripts/Server/Game/Action/ChargedShieldAction.cs]: ok

**Review by fernando-cortez — APPROVED** [review]

**Review by dwoodruffsf — DISMISSED** [review]

**Review by dwoodruffsf — APPROVED** [review]

**Review by fernando-cortez — DISMISSED** [review]

File: Assets/BossRoom/Scripts/Server/Game/Action/ChargedShieldAction.cs [hunk 2]
```diff
@@ -74,27 +74,39 @@ public override bool ShouldBecomeNonBlocking()
             return m_StoppedChargingUpTime != 0;
         }
 
+        private float GetPercentChargedUp()
```
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Server/Game/Action/ChargedShieldAction.cs:77]: Duplicated code

**Review by dwoodruffsf — APPROVED** [review]

**Review by fernando-cortez — APPROVED** [review]

**Pull request merged**